lib/src/wallet_connect/session.dart

```diff
@@ -0,0 +1,646 @@
+    };
+
+    final response = JsonRequest(math.Random().nextInt(100000), "wc_sessionUpdate", [ result ]);
+    _communicator!.publish(_remotePeerId!, response);
+    await _webSocket?.close();
+  }
+
+  void _handleUpdateSession(JsonRequest request) {
+    final param = request.params.first;
+
+    final approved = param['approved'];
+    final chainId = param['chainId'];
+    final accounts = param['accounts']?.cast<String>();
+
+    if(approved != null && !approved) {
+      _webSocket?.close();
+    }
+  }
+
+  Future<void> _handlerSendTransaction(JsonRequest request) async {
+    final descriptionJson = jsonDecode(request.params.first as String);
+    final payload = request.params.last as String;
+
+    final bytes = Encoding.fromHex(payload.replaceFirst(RegExp(r'^0x'), ""));
+    final json = base64Decode(utf8.decode(bytes));
+
+    final protoAny = Any.fromBuffer(json);
+    GeneratedMessage message = _decodeMessage(protoAny);
+
+    final description = descriptionJson['description'];
+    final address = descriptionJson['address'];
+
+    final txPairResponsePair = await _delegate?.onApproveTransaction(description, address, message);
+    if(txPairResponsePair == null) {
+      log("rejected proposed transaction");
+      _reject(request.id);
+      return;
+    }
+
+    JsonRpcResponse response;
+    if(txPairResponsePair.txResponse.code == 0) {
+     response = JsonRpcResponse.response(request.id, txPairResponsePair.asJsonString());
+    }
+    else {
+      final message =  "${txPairResponsePair.txResponse.code} ${txPairResponsePair.txResponse.codespace} ${txPairResponsePair.txResponse.info}";
+      response = JsonRpcResponse.response(request.id, <String,dynamic>{
+        "code": "${txPairResponsePair.txResponse.code}" ,
+        "message": message,
+        "value": txPairResponsePair.asJsonString()
+      });
+    }
+    _communicator!.publish(_remotePeerId!, response);
+  }
+
+  Future<void> _handleProvenanceSign(JsonRequest request) async {
+    final descriptionJson = jsonDecode(request.params.first as String);
+    final payload = request.params.last as String;
+    final bytes = Encoding.fromHex(payload.replaceFirst(RegExp(r'^0x'), ""));
+
+    final description = descriptionJson['description'];
+    final address = descriptionJson['address'];
+
+    final signedData = await _delegate?.onApproveSign(description, address, bytes);
+    if(signedData == null) {
+      log("connection response sent");
+      _reject(request.id);
+      return;
+    }
+
+    final result = Encoding.toHex(signedData);
+
+    final response = JsonRpcResponse.response(request.id, result);
+    _communicator!.publish(_remotePeerId!, response);
+  }
+
+  Future<void> _handleSessionRequest(JsonRequest request) async {
+    final param = request.params.first;
+    _remotePeerId = param['peerId'];
+    final chainId = param['chainId'];
+    final peerMeta = param['peerMeta'];
+
+    final clientMeta = ClientMeta.fromJson(peerMeta);
```

Comment:
Might be nice to have uniformity in whether calls to `fromJson` can throw or not throw.